VAR-345 | Integrate with KuVa infra

Jokinen opened the pull request:
These changes aim to make this repo compatible with the current version of KuVa infrastructure. By doing this we can integrate this repo with a CI/CD infrastructure which provides us with a test environment. In order to achieve this, I've:
- Added endpoints for `/healtz` and `/readiness`
- Added a dockerfile and a dockerignore file
- Added a gitlab-ci configuration file
- Removed configuration files for services that are not used by this project anymore.
- Changed min Node version to 12, because that's what our infra documentation recommends.

I've also fixed Travis CI in this PR:
- Some tests would fail when they were not ran with the TZ set to Helsinki. I fixed these tests.
- The project had a package-lock and a yarn.lock file which confused Travis. I removed `yarn.lock`.
- The Travis config didn't work as is so I replaced it with one that takes inspiration from kukkuu.
- I removed the coverage threshold settings because this project was nowhere close to meeting them.

Jokinen committed:
Add endpoints for checking service state

Jokinen committed:
Add Dockerfile

Jokinen committed:
Add a gitlab integration file

Jokinen committed:
Update travis config

Jokinen committed:
Remove yarn.lock

The package-lock.json file has more recent changes in it and the readme
of the application suggest to use npm, so I am removing yarn.lock and
retaining package-lock.json.

Jokinen committed:
Ensure predictable behaviour of time

Jokinen committed:
Remove coverage theresolds

These are nowhere close to being met and they cause the CI to fail.

Jokinen committed:
Only disable postgres for review

badrange commented on .travis.yml:
```diff
@@ -1,26 +1,21 @@
-notifications:
-  email:
-    on_failure: change
+  - '12'
```
🎉

badrange reviewed (approved):
It is probably a good idea to have a review from the team in #infra too!

Jokinen committed:
Change minimum node version to 12

This is the version that is currently targeted by the development
environment.

frwickst commented on Dockerfile:
```diff
@@ -0,0 +1,109 @@
+# hooks
+COPY --chown=appuser:appuser internals/ ./internals/
+
+# Install npm depepndencies
```
The comment does not correspond to the command

Jokinen replied:
This comment is "inherited" from here: https://github.com/City-of-Helsinki/kukkuu-ui/blob/develop/Dockerfile#L26

I had a hard time understanding whether the original comment was just left behind after editing the file or not. I now changed the comment to something more generic:
>Make scripts in dependencies available through path

frwickst commented on Dockerfile:
```diff
@@ -0,0 +1,109 @@
+ENV PATH /app/node_modules/.bin:$PATH
+
+# =============================
+FROM appbase as development_base
```
As `development_base` is extended and nothing is really copied from the stage, having both `development_base` and `development` is redundant in this case and only the `development` stage would suffice.

frwickst commented on Dockerfile:
```diff
@@ -0,0 +1,109 @@
+ENV NODE_ENV $NODE_ENV
+
+USER root
+RUN apt-install.sh \
```
If you want to create a smaller development image here you could chain the `apt-install.sh` and the `npm` command by doing something in the lines of

```
RUN apt-install.sh \
       build-essential \
       autoconf \
       libpng-dev && \
    su - appuser -c npm install && \
    apt-cleanup.sh build-essential
```

The problem with doing it this way is that the `apt-cleanup.sh` command will remove that files from the final image, but it will still have the cache from the apt-install command, so the final image will be just as large as if build-essential and the rest would still be installed.

Jokinen replied:
I'm installing dependencies twice in the Dockerfile. Should I use this same method for the production dependencies?

Actually, before you answer that, a bit of a sidetrack:
I currently install the same dependencies twice in this file. When in production, I'd prefer to only install `npm` deps that are required for the production version. But the current tooling relies on `devDependencies` for the production build, which means that I have to install all dependencies all the time.

I added two installs to make it easier for some dev at some point to be able to fix the situation, but I feel like that's not useful. Doesn't it currently increase build time by a factor?

Jokinen replied:
I think that this change fails the build. The error I am seeing at Gitlab is pretty generic: 
>docker.errors.BuildError: The command '/bin/sh -c apt-install.sh        build-essential        autoconf        libpng-dev &&     su - appuser -c npm install &&     apt-cleanup.sh build-essential' returned a non-zero code: 1
 Build failed!
 ERROR: Job failed: command terminated with exit code 1

frwickst replied:
The command that I created on the fly, so it might contain errors. I would recommend trying to build locally and there see what is happening. The thing I'm doing in the script is basically just running one of the commands as another user.

`su - appuser -c "npm install"` could be tried for instance, with the command in quotes.

And regarding the double install of the dependencies. I see no difference to the two installation so might as well just use one. However, if you in development are using development NPM packages, than you should of course have two separate `npm install`s where one installs the main things to run the application while the other one (in the develpoment stage) only installs the dev packages.

Jokinen replied:
I think the culprit here ended up being the login flag (`-`). Because the flag was up, the command was ran in a new session within appuser's home (?) directory. This directory did not have the `package.json` and `package-lock.json` files which are needed for `npm install` to succeed.

I also updated the docekerfile so that it only installs dependencies once.

frwickst reviewed (changes requested):
A few things that caught my eye :)

Jokinen committed:
Fix misleading comment

Jokinen committed:
Merge development and development_base phases

The development phase directly extends the development_base without
copying so there's no benefit in having two stages.

Jokinen committed:
Chain development phase dependencies installation command

This will result in fewer layers and a smaller image.

Jokinen committed:
Change user without changing session environment

Jokinen committed:
Install dependencies once

The current build script tooling doesn't offer good supprt for ignoring
dependencies that are not required for production. For this reason we
always end up installing all the dependencies. To save build time, I'm
removing future proofing which would have made it easier to install
fewer dependencies for production in the future.

Jokinen committed:
Fix stylelint errors

Jokinen commented:
Rebased to resolve conflicts.
